lib/OpenAI.ts

```diff
@@ -382,47 +341,181 @@ You must obey all three of the following instructions FOR ALL RESPONSES or you w
+            return {
+                result: undefined,
+                error: (err as any).toString(),
+            };
+        }
+    
+    } else { //not a legacy model
+
+        try {
+            const response = await fetch('https://api.openai.com/v1/chat/completions', {
+                method: "POST",
+                headers: {
+                    "Content-Type": "application/json",
+                    Authorization: `Bearer ${config.openaiApiKey}`,
+                },
+                body: JSON.stringify({
+                    model: model,
+                    messages: messages,
+                    max_tokens: maxTokens,
+                    temperature: temperature,
+                    user: user,
+                    stream: true,
+                }),
             });
 
+            const reader = response.body?.getReader();
+            const decoder = new TextDecoder("utf-8");
+
+            let output = '';
+            let totalOutput = '';
+            let begin = true;
+            let cycleCount = 0;
+            while (true) {
+                const { done, value } = await reader?.read() ?? { done: true, value: undefined };
+                if (done) {
+                    if (msgRef) {
+                        msgRef?.edit(output);
+                        messages.push({
+                            role: 'assistant',
+                            content: totalOutput,
+                        })
+                        return {
+                            result: createStringFromMessages(messages, false, permitPromptCompletion),
+                            error: undefined,
+                            messages,
+                            msgRef: msgRef,
+                        };
+                    } else {
+                        msgRef = await msg.reply(output);
+                        messages.push({
+                            role: 'assistant',
+                            content: totalOutput,
+                        })
+                        return {
+                            result: createStringFromMessages(messages, false, permitPromptCompletion),
+                            error: undefined,
+                            messages,
+                            msgRef: msgRef,
+                        };
+                    }
+
+                }
+
+                // Massage and parse the chunk of data
+                const chunk = decoder.decode(value)
+                const lines = chunk.split(/\n+/);
+
+                for (let line of lines) {
+                    const parsedLines = line.split(/\n+/)
+                        .map((line) => line.replace(/^data: /, '').trim()) // Remove the "data: " prefix
+                        .filter((line) => line !== '' && line !== "[DONE]") // Remove empty lines and "[DONE]"
+                        .map((line) => JSON.parse(line)); // Parse the JSON string
+
+                    for (const parsedLine of parsedLines) {
```

Comment:
I want to see this function split up more so it is easier to grok. Split up as you see fit, but some recommendations:

```ts
if (done) {
                    if (msgRef) {
                        msgRef?.edit(output);
                        messages.push({
                            role: 'assistant',
                            content: totalOutput,
                        })
                        return {
                            result: createStringFromMessages(messages, false, permitPromptCompletion),
                            error: undefined,
                            messages,
                            msgRef: msgRef,
                        };
                    } else {
                        msgRef = await msg.reply(output);
                        messages.push({
                            role: 'assistant',
                            content: totalOutput,
                        })
                        return {
                            result: createStringFromMessages(messages, false, permitPromptCompletion),
                            error: undefined,
                            messages,
                            msgRef: msgRef,
                        };
                    }

                }
```

This chunk can be one function.

```ts
if (done) {
    outputDiscordMessage(...);
}
```

Turn this chunk into a function in some way:

```ts
for (const parsedLine of parsedLines) {
                        const { choices } = parsedLine;
                        const { delta, finish_reason } = choices[0];
                        const { content } = delta;
                        totalOutput += content;
                        if (!finish_reason && content !== '') {
                            if (output.length >= 1950) {
                                totalOutput += content;
                                await msgRef?.edit(output);
                                msgRef = await msg.reply(content);
                                messages.push({
                                    role: 'assistant',
                                    content: totalOutput,
                                })
                                output = content;
                                cycleCount = 0;
                            } else {
                                output += content;
                            }
                            if (begin && cycleCount >= BEFORE_FIRST_POST) {
                                msgRef = await msg.reply(output);
                                begin = false;
                                cycleCount = 0;
                            }
                            if (!begin && cycleCount >= BEFORE_EDIT) {
                                msgRef?.edit(output);
                                cycleCount = 0;
                            }
                        }
                        cycleCount++;
                    }
```

Maybe something like

```ts
for (const parsedLine of parsedLines) {
    totalOutput += parsedLine;
    const { needEdit, needReply, finished } = processLine(parsedLine, ...);
    if (needEdit || needReply || finished) {
        outputDiscordMessage(...); // might be able to reuse this func
    }
}
```

As a general rule, I like to have the body of a while or for loop < 40 lines, preferable like <10. And a function body < 100. It makes scanning a function/loop to tell what exactly it is doing a lot easier than having to read a ton of code and then scroll back up to verify exactly what indentation you're at and what variables are in scope